opt: optimize left join when no rows found on right input

sgasho opened the pull request:
#### opt: optimize left join when no rows found on right input

This commit adds a new norm rule 'SimplifyLeftJoinWithZeroRowsRight',
which matches a left join with zero rows in the right input and replaces
it with a project operator that passes left rows through and null-extends
them with the columns from the right input.

Fixes #132327

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146283)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
Looks good! I just have a few nits and some test suggestions.

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146283#-:-OPgJATv6-ySNBTo0gRm:bvl100x)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/norm/general_funcs.go` line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146283#-OPgIekn3xe5oh9MgtWf:-OPgIekn3xe5oh9MgtWg:bodfipc) ([raw file](https://github.com/cockroachdb/cockroach/blob/697c3d47450d9b801497a85ddc9da2ab1a3043c3/pkg/sql/opt/norm/general_funcs.go#L166)):*
> ```Go
> // MakeNullProjections creates a projection for each column in the right input,
> // producing NULL values for each column in the right input.
> func (c *CustomFuncs) MakeNullProjections(right memo.RelExpr) memo.ProjectionsExpr {
> ```

nit: this function has a very specific purpose and is only used in one rule, so let's move it to `join_funcs.go` or `project_funcs.go`.
___
*[`pkg/sql/opt/norm/testdata/rules/join` line 2518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146283#-OPgH_6D9tbNXbt7Pn13:-OPgH_6D9tbNXbt7Pn14:b-dy2ljt) ([raw file](https://github.com/cockroachdb/cockroach/blob/697c3d47450d9b801497a85ddc9da2ab1a3043c3/pkg/sql/opt/norm/testdata/rules/join#L2518)):*
> ```
> # SimplifyLeftJoinWithZeroRowsRight
> # --------------------------------------------------
> norm expect=SimplifyLeftJoinWithZeroRowsRight
> ```

This is a good start. I'd also like to see some "no-op" rules, cases where the rule is expected _not_ to fire:
* Case with an `InnerJoin` instead of a `LeftJoin`
* Case where the right input can have more than zero rows
___
*[`pkg/sql/opt/norm/rules/join.opt` line 340 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146283#-OPgIWsj97FKtEDcdJot:-OPgIWsj97FKtEDcdJou:b-rg3zmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d0024f360fa3cdc0b64ca92f6d8adb0f65efa0a/pkg/sql/opt/norm/rules/join.opt#L340)):*
> ```
>     $left
>     (MakeNullProjections $right)
>     (PassthroughCols $left)
> ```

nit: rather than adding a new `PassthroughCols` custom function, you can use the existing `OutputCols`, which will do the same thing.


<!-- Sent from Reviewable.io -->

sgasho commented:
@DrewKimball 
Thank you for the detailed review! 
Based on your feedback, I have made the following changes to the latest commit:

* I moved 'MakeNullProjections' from 'general_funcs.go' to 'join_funcs.go'.
* I deleted 'PassthroughCols' and replaced it with 'OutputCols'.
* I also added two test cases to ensure that SimplifyLeftJoinWithZeroRightRows is not applied.

sgasho commented on pkg/sql/opt/norm/rules/join.opt:
```diff
@@ -324,6 +324,28 @@
+(LeftJoin
+    $left:*
+    $right:* & (HasZeroRows $right)
+    $on:[
+        ...
+        $item:* & ^(IsContradiction $item)
+        ...
+    ]
```
I decided to add this condition to narrow the scope of application and avoid producing diffs at the test below(at: pkg/sql/opt/norm/testdata/rules/join)

It would be better to narrow the scope first and then widen it. If a bug occurs after widening the scope, only the widened portion can be reverted.

https://github.com/cockroachdb/cockroach/blob/acd03343f136d89a9e629c538df2ae1ba2bbb2d6/pkg/sql/opt/norm/testdata/rules/join#L263-L284

DrewKimball replied:
>I decided to add this condition to narrow the scope of application and avoid producing diffs at the test below

I think you should leave out the condition, actually, and accept the diff. You can rewrite all of the optimizer tests like this: `./dev test pkg/sql/opt/... --rewrite`
^ if you do that, many test will "fail" because they don't accept the rewrite flag, but it's fine to ignore them.

sgasho replied:
@DrewKimball 
Thanks for sharing the useful command!
I left out the condition and ran the command ./dev test pkg/sql/opt/... --rewrite.

sgasho commented:
@michae2 @DrewKimball 
Could you please review this PR again? 
I refined the commit messages and ensured that Bazel Extended CI passed.

sgasho commented:
@DrewKimball 
Hi, I would like you to re-review this PR.
Thank you!

DrewKimball committed:
opt: optimize left join when no rows found on right input

This commit adds a new norm rule 'SimplifyLeftJoinWithZeroRowsRight',
which matches a left join with zero rows in the right input and replaces
it with a project operator that passes left rows through and null-extends
them with the columns from the right input.

Fixes #132327

Release note: None

DrewKimball reviewed (approved):
Nice work @sgasho, this <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. I've squashed the commits into one, adjusted a couple comments, and modified some of the existing tests to make sure they still test what they were meant to test. I'll merge this once it passes CI.

> Reviewed 4 of 4 files at r3, 1 of 3 files at r4, 2 of 4 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146283#-:-OQLO9YcClmF5M2coK4o:b-68ikrb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @sgasho)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
bors r=drewkimball
